File: 3.x.x/crates/utopia-emails/src/canonicals/generic.rs
use crate::canonicals::provider::{to_lower_case, Canonical, Provider};
use crate::EmailError;

/// Generic canonicalization (PHP `Generic`) - lowercase only, original domain kept.
#[derive(Debug, Clone, Copy, Default)]
pub struct Generic;

impl Provider for Generic {
    fn supports(&self, _domain: &str) -> bool {
        true
    }

    fn get_canonical(&self, local: &str, domain: &str) -> Result<Canonical, EmailError> {
        Ok(Canonical {
            local: to_lower_case(local),
            domain: domain.to_string(),
        })
    }

    fn get_canonical_domain(&self) -> &'static str {
        ""
    }

    fn get_supported_domains(&self) -> &'static [&'static str] {
        &[]
    }
}

File: packages/emails/src/Canonicals/Providers/Generic.php
<?php

namespace Utopia\Emails\Canonicals\Providers;

use Utopia\Emails\Canonicals\Provider;

/**
 * Generic
 *
 * Handles generic email normalization for unsupported providers
 * - Preserves all characters in local part (no subaddress or dot removal)
 * - Only converts to lowercase
 */
class Generic extends Provider
{
    public function supports(string $domain): bool
    {
        // Generic provider supports all domains
        return true;
    }

    public function getCanonical(string $local, string $domain): array
    {
        // Convert to lowercase
        $normalizedLocal = $this->toLowerCase($local);

        // Generic providers don't remove subaddresses or dots
        // Just normalize case

        return [
            'local' => $normalizedLocal,
            'domain' => $domain,
        ];
    }

    public function getCanonicalDomain(): string
    {
        // Generic provider doesn't have a canonical domain
        return '';
    }

    public function getSupportedDomains(): array
    {
        // Generic provider supports all domains
        return [];
    }
}
